### PR TITLE
Zl/chain

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
- [x] Review and merge: https://github.com/williamofockham/NetBricks/pull/69 or just let this subsume that one. 

This extends https://github.com/williamofockham/NetBricks/pull/69 for the chain test and adds the filter_map operator built with options, just like how the `iter` one works.

### CODE REVIEW

File: examples.sh [hunk 1]
```diff
@@ -11,10 +11,10 @@ export examples=(
         # examples/maglev
         # examples/sctp-test
         # examples/tcp-reconstruction
-        # examples/chain-test
         ### Runnable examples | No Tests associated
         ### =======================================
-	examples/acl-fw
+        examples/chain
+	    examples/acl-fw
```
Comment (drunkirishcoder) [inline examples.sh]: some extra spaces.

File: framework/src/operators/mod.rs [hunk 2]
```diff
@@ -160,6 +172,33 @@ pub mod tests {
...
+                } else {
+                    Ok(None)
+                }
+            });
+
+            producer.enqueue(RawPacket::from_bytes(&UDP_PACKET).unwrap());
+            producer.enqueue(RawPacket::from_bytes(&ICMPV4_PACKET).unwrap());
+
+            let p1 = batch.next().unwrap().unwrap();
+            assert_eq!(ProtocolNumbers::Udp, p1.protocol());
```
Comment (drunkirishcoder) [inline framework/src/operators/mod.rs]: should assert something you've changed in the `filter_map`.

Comment (zeeshanlakhani) [reply framework/src/operators/mod.rs]: ok

**Comment by zeeshanlakhani** [comment]
@drunkirishcoder i mean, i agree that i can blur the line, but so can filter_map in iter tools for example. I do agree like parsing within filter should be an eventual option, but I also think you may want to make changes, and filter on a check in one go.

**Comment by clinedome** [comment]
I think I generally agree with Dan that what this mostly addresses is the need to parse inside `filter`. If we could do that and build our filter predicate on data from different parts of the packet (while not mutating anything), then I think it's possible we'd feel good with just `filter` and `map` as independent operators.

**Comment by clinedome** [comment]
Same goes for `group_by` though, for that matter... both of those are currently only able to focus on a single header within the packet for decision-making, which can lead to some cumbersome combinations when trying to do fairly simple things.

**Comment by drunkirishcoder** [comment]
Yep. The `group_by` selector has the same problem. Not an easy problem to solve since the new typed packet relies very heavily on having the ownership to do basic parsing. It will need a pretty fundamental change. Could be wrong though.

As for modeling after `iter_tools`, these are definitely inspired by that. But there are also many things that don't translate well. And what makes me kinda uneasy about this comes down to `filter_map` essentially replaces both `filter` and `map` completely. Like you can easily make a case that there no need to use anything other than `filter_map`.

**Comment by zeeshanlakhani** [comment]
Sometimes I like to filter in my maps dawg. haha.

**Pull request merged**